iroh-relay/src/quic.rs

```diff
@@ -0,0 +1,301 @@
+
+            let task = tokio::task::spawn(async move {
+            let mut set = JoinSet::new();
+            debug!("waiting for connections...");
+            loop {
+                tokio::select! {
+                    biased;
+                    _ = cancel_accept_loop.cancelled() => {
+                        break;
+                    }
+                    Some(res) = set.join_next(), if !set.is_empty() => {
+                        if let Err(err) = res {
+                            if err.is_panic() {
+                                panic!("task panicked: {:#?}", err);
+                            }
+                            warn!("connection failed: {err}");
+                        }
+                    }
+                    res = endpoint.accept() => match res {
+                        Some(conn) => {
+                             debug!("accepting connection from {:?}", conn.remote_address())       ;
+                             set.spawn(async move {
+                                 let remote_addr = conn.remote_address();
+                                 let res = handle_connection(conn).await;
+                                 if let Err(ref err) = res {
+                                     warn!(remote_address = ?remote_addr, "error handling connection {err:?}")
+                                 }
+                                 res
+                             });
+                        }
+                        None => {
+                            debug!("endpoint closed");
+                            break;
+                        }
+                    }
+                }
+            }
+            endpoint
+                .close(QUIC_ADDR_DISC_CLOSE_CODE, QUIC_ADDR_DISC_CLOSE_REASON);
+            endpoint.wait_idle().await;
+            debug!("quic endpoint has been shutdown.");
+        }.instrument(info_span!("quic-endpoint")),);
+            Ok(Self {
+                bind_addr,
+                cancel,
+                handle: AbortOnDropHandle::new(task),
+            })
+        }
+
+        /// Closes the underlying QUIC endpoint and the tasks running the
+        /// QUIC connections.
+        pub fn shutdown(&self) {
+            self.cancel.cancel();
+        }
+    }
+
+    /// A handle for the Server side of QUIC address discovery.
+    ///
+    /// This does not allow access to the task but can communicate with it.
+    #[derive(Debug, Clone)]
+    pub struct ServerHandle {
+        cancel_token: CancellationToken,
+    }
+
+    impl ServerHandle {
+        /// Gracefully shut down the quic endpoint.
+        pub fn shutdown(&self) {
+            self.cancel_token.cancel()
+        }
+    }
+
+    async fn handle_connection(conn: quinn::Incoming) -> Result<()> {
+        let connection = conn.await?;
+        info!("established");
+        // wait for the client to close the connection
+        let connection_err = connection.closed().await;
```

Comment:
Do we need a timeout on this?  Or do we allow this to be connected forever?  I guess forever is fine and allows for more interesting use in the future.

Reply:
yea, i had this same question and went for the easiest path right now. Trying to understand how this could be abused, but since we don't accept any streams, it seems like the worst someone can do is open up a ton of connections and just hang around? But then they are also hanging around with a bunch of connection opens 😂 Anyway, if you don't have a strong opinion about this right now, i'd like to leave it as is.

Reply:
Agreed that opening a ton of connections is the biggest malicious use right now.  And that's covered by the generic "do we need connection limits" issue.  So agree that we can leave this like this for now.